stream.go

```diff
@@ -0,0 +1,120 @@
+package dht
+
+import (
+	"context"
+	"sync"
+
+	pbio "github.com/gogo/protobuf/io"
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	"golang.org/x/xerrors"
+)
+
+func (dht *IpfsDHT) newStream(ctx context.Context, p peer.ID) (*stream, error) {
+	s, err := dht.host.NewStream(ctx, p, dht.protocols...)
+	if err != nil {
+		return nil, xerrors.Errorf("opening stream: %w", err)
+	}
+	ps := &stream{
+		stream: s,
+		w:      newBufferedDelimitedWriter(s),
+		r:      pbio.NewDelimitedReader(s, inet.MessageSizeMax),
+		m:      make(chan chan *pb.Message, 1),
+	}
+	go func() {
+		ps.reader()
```

Comment:
what does this do?

Reply:
This reads from the stream until we can no longer do it. In particular this lets us synchronously handle stream failures without needing callbacks and cleanup code to do it via notifications, or to lazily discover errors when attempting to use a stream at a later point.